### ISSUE TITLE
Mutations sometimes blocked by "Not found column" when operating on other column

### ISSUE THREAD

**Issue by strtgbb** [ClickHouse/ClickHouse#63545]
**Describe what's wrong**

Mutations sometimes get stuck when a column in the same table is modified.
The column referred to in the related error is typically not the one the current mutation is trying to act on.

**Does it reproduce on the most recent release?**

Tested primarily on 24.3.2.23, but other versions are affected.

**Enable crash reporting**

N/A, not a crash

**How to reproduce**

I unfortunately don't have a procedure that reproduces the error reliably, but I have encountered it multiple times through different combinations of alter commands on replicated tables with minio storage.

1. Run some combination of adding, removing, renaming and clearing columns and indexes.
2. Run another command, I usually see the issue with DROP INDEX or UPDATE COLUMN, but I've also seen a similar error in SELECT.
3. Check system.mutations for errors, usually, there are none.

**Expected behaviour**

Column mutations are not blocked by mutations on other columns.

**Error message and/or stacktrace**

Some examples from system.mutations:

```sql
Row 1:
──────
database:                   default
table:                      table0_tiered
mutation_id:                0000000064
command:                    UPDATE value39 = value39 * 5 WHERE value39 < 10000
create_time:                2024-05-07 17:55:13
block_numbers.partition_id: ['0','1','2','3']
block_numbers.number:       [103,104,183,114]
parts_to_do_names:          ['2_168_168_0_182']
parts_to_do:                1
is_done:                    0
is_killed:                  0
latest_failed_part:         2_168_168_0_182
latest_fail_time:           2024-05-07 18:30:22
latest_fail_reason:         Code: 10. DB::Exception: Not found column c77071: in block key DateTime UInt32(size = 0), value1 UInt16 UInt16(size = 0), c51536 UInt16 UInt16(size = 0), value5 UInt16 UInt16(size = 0), c60760 UInt16 UInt16(size = 0), value7 UInt16 UInt16(size = 0), value8 UInt16 UInt16(size = 0), value9 UInt16 UInt16(size = 0), c54309 UInt16 UInt16(size = 0), c1518 UInt16 UInt16(size = 0), value13 UInt16 UInt16(size = 0), value14 UInt16 UInt16(size = 0), value16 UInt16 UInt16(size = 0), value17 UInt16 UInt16(size = 0), value18 UInt16 UInt16(size = 0), c60935 UInt16 UInt16(size = 0), value21 UInt16 UInt16(size = 0), c94114 UInt16 UInt16(size = 0), c89841 UInt16 UInt16(size = 0), value25 UInt16 UInt16(size = 0), value26 UInt16 UInt16(size = 0), c62108 UInt16 UInt16(size = 0), value29 UInt16 UInt16(size = 0), value31 UInt16 UInt16(size = 0), value32 UInt16 UInt16(size = 0), c4682 UInt16 UInt16(size = 0), value35 UInt16 UInt16(size = 0), value36 UInt16 UInt16(size = 0), value37 UInt16 UInt16(size = 0), value38 UInt16 UInt16(size = 0), value39 UInt16 UInt16(size = 0), value40 UInt16 UInt16(size = 0), value41 UInt16 UInt16(size = 0), value42 UInt16 UInt16(size = 0), value43 UInt16 UInt16(size = 0), value44 UInt16 UInt16(size = 0), value45 UInt16 UInt16(size = 0), c69968 UInt16 UInt16(size = 0), c59122 UInt16 UInt16(size = 0), value49 UInt16 UInt16(size = 0), c16189 UInt16 UInt16(size = 0), c59999 UInt16 UInt16(size = 0), c20410 UInt16 UInt16(size = 0), c54561 UInt16 UInt16(size = 0), c39354 UInt16 UInt16(size = 0), c95420 UInt16 UInt16(size = 0), c18017 UInt16 UInt16(size = 0). (NOT_FOUND_COLUMN_IN_BLOCK) (version 24.3.2.23 (official build))

Row 2:
──────
database:                   default
table:                      table0_tiered
mutation_id:                0000000065
command:                    DROP INDEX IF EXISTS index_23b0e7a3_value21
create_time:                2024-05-07 17:55:13
block_numbers.partition_id: ['0','1','2','3']
block_numbers.number:       [104,105,184,115]
parts_to_do_names:          ['2_168_168_0_182']
parts_to_do:                1
is_done:                    0
is_killed:                  0
latest_failed_part:         2_168_168_0_182
latest_fail_time:           2024-05-07 18:30:22
latest_fail_reason:         Code: 10. DB::Exception: Not found column c77071: in block key DateTime UInt32(size = 0), value1 UInt16 UInt16(size = 0), c51536 UInt16 UInt16(size = 0), value5 UInt16 UInt16(size = 0), c60760 UInt16 UInt16(size = 0), value7 UInt16 UInt16(size = 0), value8 UInt16 UInt16(size = 0), value9 UInt16 UInt16(size = 0), c54309 UInt16 UInt16(size = 0), c1518 UInt16 UInt16(size = 0), value13 UInt16 UInt16(size = 0), value14 UInt16 UInt16(size = 0), value16 UInt16 UInt16(size = 0), value17 UInt16 UInt16(size = 0), value18 UInt16 UInt16(size = 0), c60935 UInt16 UInt16(size = 0), value21 UInt16 UInt16(size = 0), c94114 UInt16 UInt16(size = 0), c89841 UInt16 UInt16(size = 0), value25 UInt16 UInt16(size = 0), value26 UInt16 UInt16(size = 0), c62108 UInt16 UInt16(size = 0), value29 UInt16 UInt16(size = 0), value31 UInt16 UInt16(size = 0), value32 UInt16 UInt16(size = 0), c4682 UInt16 UInt16(size = 0), value35 UInt16 UInt16(size = 0), value36 UInt16 UInt16(size = 0), value37 UInt16 UInt16(size = 0), value38 UInt16 UInt16(size = 0), value39 UInt16 UInt16(size = 0), value40 UInt16 UInt16(size = 0), value41 UInt16 UInt16(size = 0), value42 UInt16 UInt16(size = 0), value43 UInt16 UInt16(size = 0), value44 UInt16 UInt16(size = 0), value45 UInt16 UInt16(size = 0), c69968 UInt16 UInt16(size = 0), c59122 UInt16 UInt16(size = 0), value49 UInt16 UInt16(size = 0), c16189 UInt16 UInt16(size = 0), c59999 UInt16 UInt16(size = 0), c20410 UInt16 UInt16(size = 0), c54561 UInt16 UInt16(size = 0), c39354 UInt16 UInt16(size = 0), c95420 UInt16 UInt16(size = 0), c18017 UInt16 UInt16(size = 0). (NOT_FOUND_COLUMN_IN_BLOCK) (version 24.3.2.23 (official build))
```
The last operation performed on c77071 was adding and materializing a bloom filter index, it was not dropped or renamed, although similar errors do occur when a column is dropped or renamed.

**Additional context**

Logs from each instance in the cluster
[clickhouse1.zip](https://github.com/ClickHouse/ClickHouse/files/15251496/clickhouse1.zip)
[clickhouse2.zip](https://github.com/ClickHouse/ClickHouse/files/15251502/clickhouse2.zip)
[clickhouse3.zip](https://github.com/ClickHouse/ClickHouse/files/15251505/clickhouse3.zip)

These logs also include part and ttl manipulation, but those operations aren't critical to reproducing this issue.

**Comment by den-crane** [comment]
Can you please share

```
select * from system.parts where table = 'table0_tiered' and name = '2_168_168_0_182' format Vertical;

select * from system.parts_columns where table = 'table0_tiered' and name = '2_168_168_0_182' and column = 'c77071' format Vertical

**Comment by strtgbb** [comment]
I have provoked it again, lightweight delete is involved this time.

```sql
SELECT * FROM system.mutations WHERE is_done=0 INTO OUTFILE '/var/log/clickhouse-server/mutations.txt' FORMAT Vertical
Row 1:
──────
database:                   default
table:                      table0_tiered
mutation_id:                0000000037
command:                    UPDATE _row_exists = 0 WHERE (value8 % 23) = 18
create_time:                2024-05-10 17:15:36
block_numbers.partition_id: ['0','1','2','3']
block_numbers.number:       [41,43,41,42]
parts_to_do_names:          ['0_16_16_0_40','0_26_26_0_40','0_29_29_0_40','1_16_16_0_42','1_28_28_0_42','1_31_31_0_42','2_16_16_0_40','2_26_26_0_40','2_29_29_0_40','2_33_33_0_40','3_0_0_0_41','3_16_16_0_41','3_26_26_0_41','3_29_29_0_41','3_33_33_0_41']
parts_to_do:                15
is_done:                    0
is_killed:                  0
latest_failed_part:         1_16_16_0_42
latest_fail_time:           2024-05-10 17:59:34
latest_fail_reason:         Code: 10. DB::Exception: Not found column c85198: in block ...

Row 2:
──────
database:                   default
table:                      table0_tiered
mutation_id:                0000000038
command:                    DROP INDEX IF EXISTS index_3f5e2353_value6
create_time:                2024-05-10 17:15:36
block_numbers.partition_id: ['0','1','2','3']
block_numbers.number:       [42,44,42,43]
parts_to_do_names:          ['0_16_16_0_41','0_26_26_0_41','0_29_29_0_41','1_16_16_0_43','1_28_28_0_43','1_31_31_0_43','2_16_16_0_40','2_26_26_0_40','2_29_29_0_40','2_33_33_0_40','3_0_0_0_41','3_16_16_0_41','3_26_26_0_41','3_29_29_0_41','3_33_33_0_41']
parts_to_do:                15
is_done:                    0
is_killed:                  0
latest_failed_part:         2_26_26_0_40
latest_fail_time:           2024-05-10 17:59:34
latest_fail_reason:         Code: 10. DB::Exception: Not found column c85198: in block ...
```

```sql
SELECT * FROM system.parts WHERE name='1_16_16_0_42' and table='table0_tiered' INTO OUTFILE '/var/log/clickhouse-server/parts1.txt' FORMAT Vertical
Row 1:
──────
partition:                             1
name:                                  1_16_16_0_42
uuid:                                  00000000-0000-0000-0000-000000000000
part_type:                             Compact
active:                                1
marks:                                 4
rows:                                  21361
bytes_on_disk:                         2053595
data_compressed_bytes:                 1805484
data_uncompressed_bytes:               1922490
primary_key_size:                      50
marks_bytes:                           415
secondary_indices_compressed_bytes:    243788
secondary_indices_uncompressed_bytes:  0
secondary_indices_marks_bytes:         739
modification_time:                     2024-05-10 17:15:17
remove_time:                           1970-01-01 01:00:00
refcount:                              1
...
path:                                  /var/lib/clickhouse/disks/external/store/d39/d3948f30-a487-4e28-b17d-97da442a3a63/1_16_16_0_42/
...
last_removal_attempt_time:             1970-01-01 01:00:00
removal_state:                         Cleanup thread hasn't seen this part yet

SELECT * FROM system.parts WHERE name='2_26_26_0_40' and table='table0_tiered' INTO OUTFILE '/var/log/clickhouse-server/parts2.txt' FORMAT Vertical
...
```

Nothing in system.parts_columns
```sql
SELECT * FROM system.parts_columns WHERE name='1_16_16_0_42' and table='table0_tiered' and column='c85198' INTO OUTFILE '/var/log/clickhouse-server/parts_cols1.txt' FORMAT Vertical
SELECT * FROM system.parts_columns WHERE name='2_26_26_0_40' and table='table0_tiered' and column='c85198' INTO OUTFILE '/var/log/clickhouse-server/parts_cols2.txt' FORMAT Vertical
```

```sql
show create table table0_tiered into outfile '/var/log/clickhouse-server/create.txt'
CREATE TABLE default.table0_tiered\n(\n    `key` DateTime,\n    `value1` UInt16,\n    ...   `value46` UInt16,\n    `value47` UInt16,\n    `value49` UInt16,\n    `c23038` UInt16,\n    `c98531` UInt16,\n    `c49297` UInt16,\n    `c85455` UInt16,\n    `c89411` UInt16,\n    `c11688` UInt16,\n    `c85198` UInt16,\n    `c29753` UInt16,\n    INDEX index_35f7e3d0_value8 value8 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_3966a617_value15 value15 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_3cf5bb5d_value3 value3 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_3e2d5a6b_value15 value15 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_408dbb9b_value23 value23 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_41c1bbcf_value49 value49 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_45f7076f_value47 value47 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_b1319a88_value36 value36 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_4818e1d9_value33 value33 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_b2c2bdf7_value27 value27 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_0071c593_c85198 c85198 TYPE bloom_filter GRANULARITY 1,\n    INDEX index_0e4f426f_value40 value40 TYPE bloom_filter GRANULARITY 1\n)\nENGINE = ReplicatedMergeTree(\'/clickhouse/tables/table0_tiered\', \'{replica}\')\nPARTITION BY toQuarter(key) - 1\nORDER BY key\nSETTINGS storage_policy = \'tiered\', index_granularity = 8192
```

I have attached the unabridged output with the logs

[clickhouse1_logs.zip](https://github.com/ClickHouse/ClickHouse/files/15277461/clickhouse1_logs.zip)

**Comment by den-crane** [comment]
I guess you added this `c77071` column  today ?

**Comment by strtgbb** [comment]
I started a fresh cluster, `c77071` was not involved today.
There are roughly 50 columns in the table.

**Comment by den-crane** [comment]
try as a workaround/test to disable compact parts

```
alter table table0_tiered modify setting min_bytes_for_wide_part = 0;
```

**Comment by den-crane** [comment]
I see  `RENAME COLUMN IF EXISTS value30 TO c68311` 
May be c77071 is also a result of rename column `RENAME COLUMN IF EXISTS .... TO c77071` , but that mutation was failed/aborted ?

**Comment by strtgbb** [comment]
I've reviewed my logs and both `c77071` and `c85198` are the result of `ADD COLUMN IF NOT EXISTS`.
I will try wide parts, that's a good idea.

**Comment by strtgbb** [comment]
I have not seen this issue since disabling compact parts.

**Comment by den-crane** [comment]
@strtgbb As I understand you are QA and this is a result of an automation test? 
Please provide a normal reproducible example (STR), clickhouse-server.log, more information.

If you use KILL MUTATION then it's kinda expected.

**Comment by strtgbb** [comment]
Yes, this is an automated test. It takes several hours to get the results in this issue. I have unfortunately not found a faster way to reproduce.

Column 'c36509' gets dropped, and the column is gone from system.columns before I call KILL, which does not find anything to kill. I will rerun my test with no KILL at all just in case and report back.

**Comment by den-crane** [comment]
> I call KILL, which does not find anything to kill. I will rerun my test with no KILL at all just in case and report back.

Do you mean you have some logic to execute `KILL MUTATION` ?

**Comment by strtgbb** [comment]
After every subtest I record the output of
```sql
SELECT * FROM system.mutations WHERE is_done=0 AND latest_fail_reason != '' FORMAT Vertical
```
Then run
```sql
KILL MUTATION WHERE latest_fail_reason != ''
```
This is to work around the stuck mutations issue.
Since using wide parts is also a workaround, I have disabled the KILL commands to see what happens.
Maybe I should add some logic to only run KILL when the first query returns something.

**Comment by den-crane** [comment]
>After every subtest I record the output of

I don't follow. Why you need to do it? Why  your subtests  are not isolated and run against the same table.

**Comment by strtgbb** [comment]
They are not isolated because the intent is to find errors such as those described in this thread.

**Comment by strtgbb** [comment]
If I should stop testing when I encounter a stuck mutation, then what is the procedure to recover from a stuck mutation in production?

The tests are currently being rerun without kill, so that confusion will be cleared up when I have new results in a few hours.
I would be happy to dig through my logs with somebody, but at the moment they contain extra kills and are about 500MB in size.
I had success cutting down the test to isolate the stuck mutations, I will do the same for this select error.

I'm not sure if I follow your analogy. If I borrow a math machine that produces 66 every few minutes when I crank it, then I crank it really fast for really long and eventually 42 comes out, is that not worthy of investigation?

**Comment by den-crane** [comment]
>If I borrow a math machine that produces 66 every few minutes when I crank it, then I crank it really fast for really long and eventually 42 comes out, is that not worthy of investigation?

It's not about a machine, it's about your QA report. You cannot explain what you did with the machine to get these results. And because there is no STR we cannot state that machine is broken, maybe your test system is broken.

I make Clickhouse into a state "Not found column c77071: in block key DateTime" with some illegal interventions like "kill mutation" + "system kill", but it does not mean that it invalid or buggy behavior.

**Comment by den-crane** [comment]
>If I should stop testing when I encounter a stuck mutation, then what is the procedure to recover from a stuck mutation in production?

It should not happen in production. You should file a bug report with STR if you can do it. Reports without STR are not actionable.

**Comment by strtgbb** [comment]
After further digging, I have determined that despite the similar error message, this select issue is entirely unrelated to the stuck mutations issue. The select issue does not involve KILL, it is related to part manipulation. I will open a separate issue when I have a stable reproduction method.

Let's return this thread to the original topic of column and index mutations getting stuck. I will continue investigating that issue and report back when I have a better set of reproduction steps.
